pkg/controllers/jobflow/jobflow_controller.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2022 The Volcano Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jobflow
+
+import (
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	jobflowstate "volcano.sh/volcano/pkg/controllers/jobflow/state"
+
+	vcclientset "volcano.sh/apis/pkg/client/clientset/versioned"
+	versionedscheme "volcano.sh/apis/pkg/client/clientset/versioned/scheme"
+	informerfactory "volcano.sh/apis/pkg/client/informers/externalversions"
+	batchinformer "volcano.sh/apis/pkg/client/informers/externalversions/batch/v1alpha1"
+	flowinformer "volcano.sh/apis/pkg/client/informers/externalversions/flow/v1alpha1"
+	batchlister "volcano.sh/apis/pkg/client/listers/batch/v1alpha1"
+	flowlister "volcano.sh/apis/pkg/client/listers/flow/v1alpha1"
+	"volcano.sh/volcano/pkg/controllers/apis"
+	"volcano.sh/volcano/pkg/controllers/framework"
+	"volcano.sh/volcano/pkg/controllers/jobflow/state"
+)
+
+func init() {
+	framework.RegisterController(&jobflowcontroller{})
+}
+
+// jobflowcontroller the JobFlow jobflowcontroller type.
+type jobflowcontroller struct {
+	kubeClient kubernetes.Interface
+	vcClient   vcclientset.Interface
+
+	//informer
+	jobFlowInformer     flowinformer.JobFlowInformer
+	jobTemplateInformer flowinformer.JobTemplateInformer
+	jobInformer         batchinformer.JobInformer
+
+	//jobFlowLister
+	jobFlowLister flowlister.JobFlowLister
+	jobFlowSynced cache.InformerSynced
+
+	//jobTemplateLister
+	jobTemplateLister flowlister.JobTemplateLister
+	jobTemplateSynced cache.InformerSynced
+
+	//jobLister
+	jobLister batchlister.JobLister
+	jobSynced cache.InformerSynced
+
+	// JobFlow Event recorder
+	recorder record.EventRecorder
+
+	queue          workqueue.RateLimitingInterface
+	enqueueJobFlow func(req *apis.FlowRequest)
+
+	syncHandler func(req *apis.FlowRequest) error
+
+	maxRequeueNum int
+}
+
+func (j *jobflowcontroller) Name() string {
+	return "jobflow-controller"
+}
+
+func (j *jobflowcontroller) Initialize(opt *framework.ControllerOption) error {
```

Comment:
It is recommended to use `jf` as the abbreviation of jobflow to avoid conflict with job